plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,603 @@
+#define USB_REQ_TYPE_GET_MMIO	0xc0  /* bm Request type */
+#define USB_REQ_TYPE_SET_MMIO	0x40  /* bm Request type */
+
+/* bRequest */
+#define REQ_HUB_GET_MMIO	64
+#define REQ_HUB_SET_MMIO	65
+
+/* wValue*/
+#define MBOX_ACCESS		(1 << 10)
+
+/* wIndex, mailbox register offset */
+/* First 16 registers are Data[0]-Data[15] registers */
+#define MBOX_REG_METADATA	16
+#define MBOX_REG		17 /* no name? */
+
+/* mask for the MBOX_REG register that has no name */
+#define MBOX_ERROR		(1 << 6)/* of the u8 status field */
+#define MBOX_OPVALID		(1 << 7)/* of the u8 status field */
+
+#define MBOX_TIMEOUT		3000
+
+/* HUB operation OP codes */
+#define OP_NVM_WRITE		0x20
+#define OP_NVM_AUTH_WRITE	0x21
+#define OP_NVM_READ		0x22
+#define OP_NVM_SET_OFFSET	0x23
+#define OP_DROM_READ		0x24
+
+/* NVM metadata offset and length fields are in dword units */
+/* note that these won't work for DROM read */
+#define NVM_OFFSET_TO_METADATA(p) ((((p) / 4) & 0x3fffff) << 2) /* bits 23:2  */
+#define NVM_LENGTH_TO_METADATA(p) ((((p) / 4) & 0xf) << 24)     /* bits 27:24 */
+
+/* Default length for NVM READ */
+#define NVM_READ_LENGTH 	0x224
+
+/* NVM offset */
+#define NVM_VER_OFFSET_MAJOR	0xa
+#define NVM_VER_OFFSET_MINOR	0x9
+#define NVM_VID_OFFSET_MAJOR	0x221
+#define NVM_VID_OFFSET_MINOR	0x220
+#define NVM_PID_OFFSET_MAJOR	0x223
+#define NVM_PID_OFFSET_MINOR	0x222
+
+struct mbox_regx {
+	guint16		opcode;
+	guint8		rsvd;
+	guint8		status;
+} __attribute__ ((packed));
+
+struct _FuDellDockUsb4 {
+	FuUsbDevice	parent_instance;
+	guint		blocksz;
+	guint8		intf_nr;
+};
+
+G_DEFINE_TYPE (FuDellDockUsb4, fu_dell_dock_usb4, FU_TYPE_USB_DEVICE)
+
+/* write a uint32_t to a little endian ordered uint8_t array */
+static void
+fu_dell_dock_usb4_write_u32to_buf (guint32 val, guint8 *buf)
+{
+	buf[0] = (val >> 0) & 0xff;
+	buf[1] = (val >> 8) & 0xff;
+	buf[2] = (val >> 16) & 0xff;
+	buf[3] = (val >> 24) & 0xff;
+}
+
+/* wIndex contains the hub register offset, value BIT[10] is "access to
+ * mailbox", rest of values are vendor specific or rsvd  */
+static gboolean
+fu_dell_dock_usb4_hub_get_mmio (FuDevice *device,
+			        guint16 mbox_reg,
+			        guchar *buf,
+			        GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	struct mbox_regx *regx;
+
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    REQ_HUB_GET_MMIO, /* request */
+					    MBOX_ACCESS,      /* value */
+					    mbox_reg,	      /* index */
+					    (guchar *) buf,   /* data */
+					    4,		      /* length */
+					    NULL,	      /* actual length */
+					    MBOX_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error,
+				"GET_MMIO failed to set control on mbox register index [0x%x]:",
+				mbox_reg);
+		return FALSE;
+	}
+	/* verify status for GET_MMIO via hub mailbox register */
+	if (mbox_reg == MBOX_REG) {
```

Comment:
if it's not this, should it be an error?

Reply:
yes error for reading data registers, the status bits only for specific mailbox register.